Sources/Fuzzilli/Execution/REPRL.swift

```diff
@@ -20,71 +20,89 @@ import libreprl
-    
+
     /// Number of script executions since start of child process
     private var execsSinceReset = 0
-    
+
     /// Number of execution failures since the last successfully executed program
     private var recentlyFailedExecutions = 0
-    
+
     private var reprlContext: OpaquePointer? = nil
-    
+
+    /// Buffer to hold scripts, this lets us debug issues that arise if
+    /// previous scripts corrupted any state which is discovered in
+    /// future executions.
+    private var scriptBuffer: [String]? = nil
+
     public init(executable: String, processArguments: [String], processEnvironment: [String: String]) {
         self.processArguments = [executable] + processArguments
         super.init(name: "REPRL")
-        
+
         for (key, value) in processEnvironment {
             env.append(key + "=" + value)
         }
     }
-    
+
     override func initialize() {
         reprlContext = libreprl.reprl_create_context()
         if reprlContext == nil {
             logger.fatal("Failed to create REPRL context")
         }
 
+        if fuzzer.config.diagnostics {
+            self.scriptBuffer = [String]()
+            self.scriptBuffer!.reserveCapacity(maxExecsBeforeRespawn)
+        }
+
         let argv = convertToCArray(processArguments)
         let envp = convertToCArray(env)
-        
-        if reprl_initialize_context(reprlContext, argv, envp, /* capture_stdout: */ 0, /* capture stderr: */ 1) != 0 {
+
+        if reprl_initialize_context(reprlContext, argv, envp, /* capture_stdout: */ fuzzer.config.diagnostics ? 1 : 0, /* capture stderr: */ 1) != 0 {
             logger.fatal("Failed to initialize REPRL context: \(String(cString: reprl_get_last_error(reprlContext)))")
         }
-        
+
         freeCArray(argv, numElems: processArguments.count)
         freeCArray(envp, numElems: env.count)
 
         fuzzer.registerEventListener(for: fuzzer.events.Shutdown) {
             reprl_destroy_context(self.reprlContext)
         }
     }
-    
+
     public func setEnvironmentVariable(_ key: String, to value: String) {
         env.append(key + "=" + value)
     }
-    
+
     public func run(_ script: String, withTimeout timeout: UInt32) -> Execution {
+        // Log the current script into the buffer if diagnostics are enabled.
+        if fuzzer.config.diagnostics {
+            self.scriptBuffer!.append(script)
+        }
+
         let execution = REPRLExecution(in: reprlContext)
-        
+
         guard script.count <= REPRL_MAX_DATA_SIZE else {
             logger.error("Script too large to execute. Assuming timeout...")
             execution.outcome = .timedOut
             return execution
         }
-        
+
         execsSinceReset += 1
         var freshInstance: Int32 = 0
         if execsSinceReset > maxExecsBeforeRespawn {
             freshInstance = 1
             execsSinceReset = 0
+            if fuzzer.config.diagnostics {
+                scriptBuffer!.removeAll()
```

Comment:
With `scriptBuffer` being a string, this can use https://developer.apple.com/documentation/swift/string/1641517-removeall and do `keepingCapacity: true`